pkg/availability/multisigcollector/parts/certcreation/certcreation.go

```diff
@@ -95,17 +94,18 @@ func IncludeCreatingCertificates(
 	// When the mempool provides a batch, compute its ID.
 	mempooldsl.UponNewBatch(m, func(txIDs []tt.TxID, txs []*requestpbtypes.Request, context *requestBatchFromMempoolContext) error {
 		if len(txs) == 0 {
+			// If the batch is empty, immediately stop it and respond to all pending requests with an empty certificate.
 			delete(state.Certificates, context.reqID)
-			if len(state.RequestStates) > 0 { // do not do work for the empty batch if there is not even a request to answer
-				sendIfReady(m, &state, params, true)
-				if len(state.Certificates) == 0 {
-					apbdsl.ComputeCert(m, mc.Self)
-				}
-			}
-			return nil
+			respond(m, &state, []*mscpbtypes.Cert{})
+			// Note that a creation of another certificate is NOT started here.
+			// It will only be triggered by another certificate request (UponRequestCert).
+			// This creates a slight inconvenience that, if there is an empty batch no certificates will be pre-computed
+			// for the next request, and the next request will have to wait until the availability protocol finishes.
+			// TODO: A remedy would be to start a timer here, that would trigger a delayed ComputeCert event
+			//   (it's a one-liner here, but the corresponding delay parameter needs to be introduced in the params).
```

Comment:
1. In the previous implementation, the loop would also not continue, because the repetition was guarded by the `len(state.RequestStates) > 0` condition. Thus the loop would run once and then stop.
2. I think there is a problem with such an infinite loop though. While it doesn't bloat memory, it's busy-waiting and burns CPU at the maximum rate, since it _always_ generates something for a CPU to do. As a side effect, it also breaks the simulated time in tests, since it generates an infinite causal chain of events.

If we do put a timer here (but I think that even the current solution is ok for now), it can be a very very short one (even tens of milliseconds). This would give us basically the same advantage as a busy-wait loop, without killing the equivalent of a whole CPU core.

Reply:
Fair enough! I reckon best would be some code at the mempool that specifically only returns once it has a batch, paired with a timer at the mempool to return an empty batch otherwise (perhaps with a flag set by the requester, i.e. the MSC, to have a small timer or a big one). Something of this sort, but for now we can leave it as is.